common/src/main/java/fudge/notenoughcrashes/utils/CrashLogUpload.java

```diff
@@ -46,33 +49,109 @@ public GistFile(String content) {
         }
     }
 
-    public static String upload(String text) throws IOException {
-        // Don't give a fuck about your preferences
-        return uploadToGist(text);
-//        ModConfig.CrashLogUploadType type = ModConfig.instance().uploadCrashLogTo;
-//        if (type == ModConfig.CrashLogUploadType.GIST) {
-//            return uploadToGist(text);
-//        } else {
-//            return uploadToHaste(text);
-//        }
+    private static ArrayList<ModConfig.CrashLogUploadType> fallBackTypes;
+
+    private static ArrayList<ModConfig.CrashLogUploadType> getFallBackTypes(Boolean force) {
+        if (fallBackTypes != null && !force) {
+            return fallBackTypes;
+        }
+        fallBackTypes = new ArrayList<ModConfig.CrashLogUploadType>();
+        Collections.addAll(fallBackTypes, ModConfig.CrashLogUploadType.values());
+        fallBackTypes.sort(new CrashLogUploadTypeComparator());
+        fallBackTypes.removeIf(fallbackType -> fallbackType.getPriority() < 0);
+
+        return  fallBackTypes;
+    }
+    private static ArrayList<ModConfig.CrashLogUploadType> getFallBackTypes() {
+        return getFallBackTypes(false);
+    }
+
+    private static class CrashLogUploadTypeComparator implements Comparator<ModConfig.CrashLogUploadType> {
+        @Override
+        public int compare(ModConfig.CrashLogUploadType o1, ModConfig.CrashLogUploadType o2) {
+            return Integer.compare(o1.getPriority(),o2.getPriority());
+        }
+    }
+
+    public static String upload(String text) throws IOException { return upload(text,false); }
+
+    public static String upload(String text, Boolean fallBack) throws IOException {
+        String URL = "";
+        ModConfig.CrashLogUploadType uploadType;
+        if (fallBack && getFallBackTypes().isEmpty()) {
+            throw new IOException("no valid fallbacks");
+        } else if (fallBack) {
+            uploadType = getFallBackTypes().remove(0);
+        } else { uploadType = ModConfig.instance().uploadCrashLogTo; }
+
+
+    try {
+
+        switch (uploadType) {
+            case GIST:
+            default:
+                String GISTuploadKey = ModConfig.instance().GISTUploadKey;
+
+                if (GISTuploadKey == "" || fallBack) {
+                    GISTuploadKey = GIST_ACCESS_TOKEN;
+                }
+                URL = uploadToGist(text, GISTuploadKey);
+                break;
+            case HASTE:
+                String hasteUrl = ModConfig.instance().HASTEUrl;
+                if (hasteUrl == "" || fallBack) {
```

Comment:
why `|| fallback`  here?

Reply:
so it defaults to to the default url (official instance) when falling back, since we cant trust user input